tensorflow/core/kernels/data/map_dataset_op_test.cc

```diff
@@ -0,0 +1,363 @@
+
+    TF_CHECK_OK(
+        CreateOpKernelContext(map_kernel_.get(), &map_inputs_, &map_context_));
+    TF_CHECK_OK(CheckOpKernelInput(*map_kernel_, map_inputs_));
+    DatasetBase* map_dataset;
+    TF_CHECK_OK(
+        CreateDataset(map_kernel_.get(), map_context_.get(), &map_dataset));
+    return map_dataset;
+  }
+
+ private:
+  // Creates a new MapDataset op kernel. The `input_dataset` parameter should be
+  // same with the node name of the input dataset for the method
+  // `CreateMapDataset()`.
+  template <typename T>
+  std::unique_ptr<OpKernel> CreateMapDatasetOpKernel(
+      const string& input_dataset, const string& func_name) {
+    DataTypeVector map_output_dtypes({tensorflow::DataTypeToEnum<T>::value});
+    gtl::ArraySlice<TensorShape> map_output_shapes({{}});
+    FunctionDefHelper::AttrValueWrapper func =
+        FunctionDefHelper::FunctionRef(func_name, {{"T", DT_INT64}});
+
+    map_node_def_ = test::function::NDef(kNodeName, kOpName, {input_dataset},
+                                         {{"f", func},
+                                          {"Targuments", {}},
+                                          {"output_shapes", map_output_shapes},
+                                          {"output_types", map_output_dtypes},
+                                          {"use_inter_op_parallelism", true},
+                                          {"preserve_cardinality", false}});
+    std::unique_ptr<OpKernel> map_op_kernel;
+    TF_CHECK_OK(CreateOpKernel(map_node_def_, &map_op_kernel));
+    return map_op_kernel;
+  }
+
+ protected:
+  DatasetBase* range_dataset_;
+  DatasetBase* map_dataset_;
+  std::unique_ptr<OpKernel> map_kernel_;
+  std::unique_ptr<OpKernelContext> map_context_;
+  std::unique_ptr<IteratorContext> iterator_context_;
+  std::unique_ptr<IteratorBase> iterator_;
+  std::unique_ptr<SerializationContext> serialization_context_;
+
+ private:
+  NodeDef map_node_def_;
+  gtl::InlinedVector<TensorValue, 4> map_inputs_;
+};
+
+const char* MapDatasetOpTest::kNodeName = "map_dataset";
+const char* MapDatasetOpTest::kOpName = "MapDataset";
+
+struct DatasetGetNextTest
+    : MapDatasetOpTest,
+      ::testing::WithParamInterface<
+          std::tuple<int64, int64, int64, string, std::vector<int64>,
+                     std::vector<FunctionDef>>> {};
+
+TEST_P(DatasetGetNextTest, GetNext) {
+  int thread_num = 2, cpu_num = 2;
+  int64 start = std::get<0>(GetParam());
+  int64 end = std::get<1>(GetParam());
+  int64 step = std::get<2>(GetParam());
+  string func_name = std::get<3>(GetParam());
+  std::vector<int64> expected_values = std::get<4>(GetParam());
+  std::vector<FunctionDef> func_lib = std::get<5>(GetParam());
+
+  TF_ASSERT_OK(InitThreadPool(thread_num));
+  TF_ASSERT_OK(InitFunctionLibraryRuntime(func_lib, cpu_num));
+  TF_ASSERT_OK(
+      CreateRangeDataset<int64>(start, end, step, "range", &range_dataset_));
+  map_dataset_ = CreateMapDataset(range_dataset_, func_name);
```

Comment:
After changing the `CreateRangeDataset and `CreateMapDataset` signatures, this should look as follows:

```
DatasetBase* range_dataset;
TF_ASSERT_OK(
      CreateRangeDataset<int64>(start, end, step, "range", &range_dataset));
core::ScopedUnref scoped_unref(range_dataset);
DatasetBase* map_dataset;
TF_ASSERT_OK(CreateMapDataset(range_dataset, func_name, &map_dataset)
core::ScopedUnref scoped_unref(map_dataset);
```